### PR TITLE
feat(designer): File picker

### PR DESCRIPTION
**Pull request by Eric-B-Wu**
filepicker2

### CODE REVIEW

**Review by hartra344 — APPROVED** [review]

File: libs/designer-ui/src/lib/picker/filepickereditor.tsx [hunk 1]
```diff
@@ -38,35 +38,67 @@ const calloutProps = { gapSpace: 0 };
 export const FilePickerEditor = ({
   initialValue,
   isLoading = false,
+  type,
+  items,
+  displayValue,
+  // fileFilters,
```
Comment (preetriti1) [inline libs/designer-ui/src/lib/picker/filepickereditor.tsx:44]: _am assuming the file filtering is coming soon?_

File: libs/designer-ui/src/lib/picker/filepickereditor.tsx [hunk 2]
```diff
@@ -38,35 +38,67 @@ const calloutProps = { gapSpace: 0 };
...
+
+  useEffect(() => {
+    if (displayValue) {
+      setPickerDisplayValue([{ id: guid(), value: displayValue, type: ValueSegmentType.LITERAL }]);
+    }
+  }, [displayValue]);
 
-  const { onFolderNavigation: onFolderNavigated } = pickerCallbacks;
+  const { onFolderNavigation, getFileSourceName, getDisplayValueFromSelectedItem, getValueFromSelectedItem } = pickerCallbacks;
+  const fileSourceName = getFileSourceName?.();
```
Comment (preetriti1) [inline libs/designer-ui/src/lib/picker/filepickereditor.tsx:66]: All callbacks are required so getFileSourceName need not have ? and similar to other callbacks as well in this file.

File: libs/designer-ui/src/lib/picker/filepickereditor.tsx [hunk 3]
```diff
@@ -38,35 +38,67 @@ const calloutProps = { gapSpace: 0 };
...
+  const onFolderNavigated = (selectedItem: any) => {
+    onFolderNavigation?.(selectedItem);
+    const displayValue = getDisplayValueFromSelectedItem?.(selectedItem);
+    setTitleSegments([...titleSegments, { text: displayValue, key: displayValue, onClick: () => onFolderNavigated(selectedItem) }]);
+  };
+
+  const onFolderSelected = (selectedItem: any) => {
+    if (showPicker) {
+      setSelectedItem(selectedItem);
+      setPickerDisplayValue([{ id: guid(), value: selectedItem.Path, type: ValueSegmentType.LITERAL }]);
```
Comment (preetriti1) [inline libs/designer-ui/src/lib/picker/filepickereditor.tsx:87]: Why are we defaulting to Path here? you should be calling getDisplayValueFromSelectedItem to set the value no?

File: libs/designer-ui/src/lib/picker/filepickereditor.tsx [hunk 4]
```diff
@@ -38,35 +38,67 @@ const calloutProps = { gapSpace: 0 };
...
-    setValue(newValue);
-    // TODO: Make sure to update displayValue and selectedItem in viewModel.
-    onChange?.({ value: newValue, viewModel: { hideErrorMessage: true } });
+  const onFolderNavigated = (selectedItem: any) => {
+    onFolderNavigation?.(selectedItem);
+    const displayValue = getDisplayValueFromSelectedItem?.(selectedItem);
+    setTitleSegments([...titleSegments, { text: displayValue, key: displayValue, onClick: () => onFolderNavigated(selectedItem) }]);
+  };
+
+  const onFolderSelected = (selectedItem: any) => {
```
Comment (preetriti1) [inline libs/designer-ui/src/lib/picker/filepickereditor.tsx:84]: I am guessing this is for both file/folder selection right?

File: libs/designer-ui/src/lib/picker/picker.tsx [hunk 5]
```diff
@@ -24,16 +24,32 @@ export interface PickerProps {
   currentPathSegments: IBreadcrumbItem[];
   files: FileItem[];
   loadingFiles?: boolean;
+  errorDetails?: { message: string };
   onCancel?: () => void;
   handleFolderNavigation?: (selectedObject: any) => void;
+  handleFolderSelected?: (selectedObject: any) => void;
```
Comment (preetriti1) [inline libs/designer-ui/src/lib/picker/picker.tsx:30]: It is better to call this as handleItemSelected to make it generic for file/folder

File: libs/designer-ui/src/lib/picker/picker.tsx [hunk 6]
```diff
@@ -58,14 +74,25 @@ export const Picker = ({ visible, onCancel, anchorId, currentPathSegments, files
...
+      ) : files.length > 0 ? (
         <ul className="msla-picker-items">
-          {files.map((file) => (
-            <li key={file.text} className="msla-item">
-              <PickerItem text={file.text} type={file.type} onNavigation={file.onNavigation} onSelect={file.onSelect} value={file.value} />
+          {files.map((file, i) => (
+            <li key={i} className="msla-item">
+              <PickerItem
+                displayName={file.displayName}
+                mediaType={file.mediaType}
```
Comment (preetriti1) [inline libs/designer-ui/src/lib/picker/picker.tsx:83]: We need to filter based on mediaType from teh fileFilters props

File: libs/designer-ui/src/lib/picker/pickerItem.tsx [hunk 7]
```diff
@@ -13,50 +18,49 @@ const iconInfo = {
...
   value: any;
-  onSelect?: () => void;
-  onNavigation?: () => void;
+  onSelect?: (selectedObject: any) => void;
+  onNavigation?: (selectedObject: any) => void;
 }
 
 export const PickerItem = (props: FileItem) => {
-  return props.type === FOLDER_TYPE ? <FolderItem {...props} /> : <FileItem {...props} />;
+  return props.mediaType === PickerItemType.FOLDER || props.isParent ? <FolderItem {...props} /> : <FileItem {...props} />;
```
Comment (preetriti1) [inline libs/designer-ui/src/lib/picker/pickerItem.tsx:49]: mediaType == folder? these normally contain the extension, so if you want to check for folder please only rely on props.isParent, same for line:62

File: libs/designer/src/lib/core/utils/parameters/dynamicdata.ts [hunk 8]
```diff
@@ -151,7 +151,10 @@ export async function getDynamicSchema(
 
       switch (dynamicState?.extension?.builtInOperation) {
         case 'getVariableSchema':
-          schema = { type: getSwaggerTypeFromVariableType(operationParameters['type']?.toLowerCase() ?? 'boolean') };
+          schema = {
+            type: getSwaggerTypeFromVariableType(operationParameters['type']?.toLowerCase() ?? 'boolean'),
+            enum: getSwaggerEnumFromVariableType(operationParameters['type']?.toLowerCase() ?? 'boolean'),
```
Comment (preetriti1) [inline libs/designer/src/lib/core/utils/parameters/dynamicdata.ts:156]: Why is this change needed?

File: libs/designer/src/lib/core/utils/parameters/dynamicdata.ts [hunk 9]
```diff
@@ -295,8 +298,8 @@ export async function getFolderItems(
     const connectionId = connectionReference?.connection.id as string;
     const { operationId, parameters: referenceParameters } = selectedValue ? browse : open;
     const pickerParameters = Object.keys(referenceParameters ?? {}).reduce((result: Record<string, any>, paramKey: string) => {
-      if (referenceParameters?.[paramKey]?.selectedItemValuePath) {
-        return { ...result, [paramKey]: getPropertyValue(selectedValue, referenceParameters[paramKey].selectedItemValuePath) };
+      if (referenceParameters?.[paramKey]?.['value-property']) {
+        return { ...result, [paramKey]: getPropertyValue(selectedValue, referenceParameters[paramKey]['value-property']) };
```
Comment (preetriti1) [inline libs/designer/src/lib/core/utils/parameters/dynamicdata.ts:302]: The reference parameter can have either of these properties, so please check first for selectedItemValuePath and then for value-property
for backcompatibility

File: libs/designer/src/lib/core/utils/parameters/dynamicdata.ts [hunk 10]
```diff
@@ -663,6 +666,14 @@ function getSwaggerTypeFromVariableType(variableType: string): string | undefine
...
   }
 }
+function getSwaggerEnumFromVariableType(variableType: string): boolean[] | undefined {
+  switch (variableType) {
+    case 'boolean':
+      return [true, false];
+    default:
+      return undefined;
+  }
+}
```
Comment (preetriti1) [inline libs/designer/src/lib/core/utils/parameters/dynamicdata.ts:676]: I dont think this is needed for file/folder picker, can you remove this change

File: libs/designer/src/lib/ui/panel/panelTabs/parametersTab/index.tsx [hunk 11]
```diff
@@ -136,8 +136,8 @@ const ParameterSection = ({
       const { viewModel } = newState;
       const parameter = nodeInputs.parameterGroups[group.id].parameters.find((param: any) => param.id === id);
       if (
-        (parameter?.type === constants.SWAGGER.TYPE.BOOLEAN && value.length === 1 && value[0]?.value === 'True') ||
-        value[0]?.value === 'False'
+        (parameter?.type === constants.SWAGGER.TYPE.BOOLEAN && value.length === 1 && value[0]?.value?.toLowerCase() === 'true') ||
+        value[0]?.value?.toLowerCase() === 'false'
```
Comment (preetriti1) [inline libs/designer/src/lib/ui/panel/panelTabs/parametersTab/index.tsx:140]: this too, remove

File: libs/parsers/src/lib/models/operation.ts [hunk 12]
```diff
@@ -99,7 +99,7 @@ interface DynamicTreePickerInfo {
   parameters?: Record<
     string,
     {
-      selectedItemValuePath: string;
```
Comment (preetriti1) [inline libs/parsers/src/lib/models/operation.ts:102]: Do not remove this, this is set is manifest based operation so this is the new property connector will be using, just add the value-property one.

**Review by preetriti1 — CHANGES_REQUESTED** [review]

File: libs/designer/src/lib/core/utils/parameters/dynamicdata.ts (hunk 8 above)
Comment (Eric-B-Wu) [reply libs/designer/src/lib/core/utils/parameters/dynamicdata.ts:156]: This is a separate Fix, #2023 I changed it so it would get the enum values, but this was also getting caught on initialize variable boolean to have the options yes/no/ ' ', instead of true/false

File: libs/designer/src/lib/ui/panel/panelTabs/parametersTab/index.tsx (hunk 11 above)
Comment (Eric-B-Wu) [reply libs/designer/src/lib/ui/panel/panelTabs/parametersTab/index.tsx:140]: This is needed for #2012

File: libs/designer-ui/src/lib/picker/filepickereditor.tsx [hunk 13]
```diff
@@ -75,41 +107,51 @@ export const FilePickerEditor = ({
...
-  });
+const filterItems = (items?: TreeDynamicValue[], type?: string, fileFilters?: string[]): TreeDynamicValue[] => {
+  if (!items || items.length === 0) return [];
+  let returnItems = items;
+  if (type === PickerItemType.FOLDER) {
+    returnItems = items.filter((item) => item.isParent);
+  }
+  if (fileFilters && fileFilters.length > 0) {
+    returnItems = returnItems.filter((item) => {
+      return fileFilters.includes(item.mediaType ?? '');
```
Comment (preetriti1) [inline libs/designer-ui/src/lib/picker/filepickereditor.tsx:147]: Can you change this to fileFilters.some(filter => equals(filter, item.mediaType))

File: libs/designer/src/lib/core/utils/parameters/dynamicdata.ts (hunk 10 above)
Comment (preetriti1) [reply libs/designer/src/lib/core/utils/parameters/dynamicdata.ts:681]: Lets move it in separate PR, i need to understand that bug and review it separately.

File: libs/designer/src/lib/ui/panel/panelTabs/parametersTab/index.tsx (hunk 11 above)
Comment (preetriti1) [reply libs/designer/src/lib/ui/panel/panelTabs/parametersTab/index.tsx:144]: Still separate PR, this PR is fixing two other bugs which is not convernient if we want to revert

**Comment by Eric-B-Wu** [comment]
fix filepicker

File: libs/designer-ui/src/lib/picker/filepickereditor.tsx [hunk 14]
```diff
@@ -38,35 +40,71 @@ const calloutProps = { gapSpace: 0 };
...
+      editorBlur?.({
+        value: editorDisplayValue,
+        viewModel: { displayValue: undefined, selectedItem: undefined },
+      });
+    }
+  };
+
+  const clearPickerInfo = () => {
+    setSelectedItem(undefined);
+    setPickerDisplayValue([]);
```
Comment (preetriti1) [inline libs/designer-ui/src/lib/picker/filepickereditor.tsx:107]: Should you have it as empty value in a literal segment of valueSegment[]

File: libs/designer-ui/src/lib/picker/plugins/EditorValueChange.tsx [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+interface ChangeProps {
+  pickerDisplayValue: ValueSegment[];
+  setEditorDisplayValue: (newVal: ValueSegment[]) => void;
+  clearPickerInfo: () => void;
+}
+
+export const EditorValueChange = ({ pickerDisplayValue, setEditorDisplayValue, clearPickerInfo }: ChangeProps) => {
+  const onChange = (editorState: EditorState) => {
+    console.log('EditorValueChange');
```
Comment (preetriti1) [inline libs/designer-ui/src/lib/picker/plugins/EditorValueChange.tsx:15]: Remove console logging before checkin

**Review by preetriti1 — APPROVED** [review]
